Fix behavior of scrollspy with nested navigation

ddevault opened the pull request:
Consider a nav like this with scrollspy enabled:

```
<ul class="nav">

    ...

    <li class="nav-item">
        <a class="nav-link" href="#link-2">Link 2</a>
        <ul class="nav">
            <li class="nav-item">
                <a href="#link-2-a" class="nav-link">Link 2-A</a>
            </li>
            <li class="nav-item">
                <a href="#link-2-b" class="nav-link">Link 2-B</a>
            </li>
        </ul>
    </li>

    ...

</ul>
```

When scrolling without this patch applied and any of these links come
into view, they will all receive .active. With this patch applied, the
active link and links above it in the tree will receive .active. This
matches the behavior of Bootstrap 3.

Before:

![](https://sr.ht/BRIy.png)

After:

![](https://sr.ht/VorE.png)

Commit:
Fix behavior of scrollspy with nested navigation

Consider a nav like this with scrollspy enabled:

    <ul class="nav">

        ...

        <li class="nav-item">
            <a class="nav-link" href="#link-2">Link 2</a>
            <ul class="nav">
                <li class="nav-item">
                    <a href="#link-2-a" class="nav-link">Link 2-A</a>
                </li>
                <li class="nav-item">
                    <a href="#link-2-b" class="nav-link">Link 2-B</a>
                </li>
            </ul>
        </li>

        ...

    </ul>

When scrolling without this patch applied and any of these links come
into view, they will all receive .active. With this patch applied, the
active link and links above it in the tree will receive .active. This
matches the behavior of Bootstrap 3.

Before:

![](https://sr.ht/BRIy.png)

After:

![](https://sr.ht/VorE.png)

cvrebert commented:
@SirCmpwn This breaks some unit tests; see the Travis build output. Use `grunt dist-js test-js` to run the unit tests locally.

cvrebert commented:
CC: @fat on account of the `// todo (fat) this is kinda sucks…` comment in this section of code

SirCmpwn commented:
Thanks, working on it now.

Commit:
Fix scrollspy unit tests for 8a8d6b4

The error here is that the <li>s involved in the test did not have
.nav-item set. The scrollspy plugin explicitly states that Bootstrap
navs are the only supported target, and the base nav includes .nav-item
on all <li>s. Therefore, the test is incorrect and scrollspy is correct.

SirCmpwn commented:
Thanks for waiting. I've added a commit that addresses this issue.

SirCmpwn commented:
Bump.

SirCmpwn commented:
@cvrebert mind taking a second look at this?

displague commented:
I need this fix, bump

akalata commented:
Is this a dupe of https://github.com/twbs/bootstrap/pull/17770?

SirCmpwn commented:
Looks like it, yes.

akalata commented:
I tested the scrollspy.js fix manually, works for me thanks.

mdo commented:
This, in affect, then requires the use of `.nav-item` around `.nav-link`, which is one of the things we were trying to be more flexible with in v4. Closing for now, thanks though!
